pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -0,0 +1,779 @@
+                        next == null ? "" : next);
+            }
+        } else {
+            log.error("{} failed to handle {} event for serviceUnit:{}, cur:{}, next:{}",
+                    lookupServiceAddress, getLogEventTag(data), serviceUnit,
+                    data == null ? "" : data,
+                    next == null ? "" : next,
+                    e);
+        }
+    }
+
+    private void handleOwnEvent(String serviceUnit, ServiceUnitStateData data) {
+        var getOwnerRequest = getOwnerRequests.remove(serviceUnit);
+        if (getOwnerRequest != null) {
+            getOwnerRequest.complete(data.broker());
+        }
+        if (isTargetBroker(data.broker())) {
+            log(null, serviceUnit, data, null);
+        }
+
+    }
+
+    private void handleAssignEvent(String serviceUnit, ServiceUnitStateData data) {
+        deferGetOwnerRequest(serviceUnit);
+        if (isTargetBroker(data.broker())) {
+            ServiceUnitStateData next = new ServiceUnitStateData(
+                    isTransferCommand(data) ? Released : Owned, data.broker(), data.sourceBroker());
+            pubAsync(serviceUnit, next)
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, next));
+        }
+    }
+
+    private void handleReleaseEvent(String serviceUnit, ServiceUnitStateData data) {
+
+        if (isTargetBroker(data.sourceBroker())) {
+            ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker(), data.sourceBroker());
+            // TODO: when close, pass message to clients to connect to the new broker
+            closeServiceUnit(serviceUnit)
+                    .thenCompose(__ -> pubAsync(serviceUnit, next))
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, next));
+        }
+    }
+
+    private void handleSplitEvent(String serviceUnit, ServiceUnitStateData data) {
+        if (isTargetBroker(data.broker())) {
+            splitServiceUnit(serviceUnit)
+                    .thenCompose(__ -> tombstoneAsync(serviceUnit))
+                    .whenComplete((__, e) -> log(e, serviceUnit, data, null));
+        }
+    }
+
+    private void handleTombstoneEvent(String serviceUnit) {
+        closeServiceUnit(serviceUnit)
+                .thenAccept(__ -> {
+                    var request = getOwnerRequests.remove(serviceUnit);
+                    if (request != null) {
+                        request.completeExceptionally(new IllegalStateException("The ownership has been unloaded. "
+                                + "No owner is found for serviceUnit: " + serviceUnit));
+                    }
+                })
+                .whenComplete((__, e) -> log(e, serviceUnit, null, null));
+    }
+
+    private CompletableFuture<MessageId> pubAsync(String serviceUnit, ServiceUnitStateData data) {
+        CompletableFuture<MessageId> future = new CompletableFuture<>();
+        try {
+            outstandingPubMessages.acquire();
```

Comment:
blocking in a method that returns a CompletableFuture is usually a code smell and it leads to performance problems.
BUT At the moment I am not aware of a better way to do this in Pulsar codebase

Reply:
The intention was to limit the `pubAsync` concurrency from this channel not to exhaust the broker threads.

I removed this semaphore.